File: rpc/cluster/DatabaseManagerClusterRPC.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package grakn.client.rpc.cluster;

import grakn.client.GraknClient;
import grakn.client.common.exception.GraknClientException;
import grakn.client.rpc.DatabaseManagerRPC;
import grakn.client.rpc.DatabaseRPC;
import grakn.protocol.cluster.DatabaseProto;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static grakn.client.common.exception.ErrorMessage.Client.CLUSTER_ALL_NODES_FAILED;

public class DatabaseManagerClusterRPC implements GraknClient.DatabaseManager.Cluster {
    private final Map<ServerAddress, DatabaseManagerRPC> databaseManagers;
    private final ClientClusterRPC client;

    public DatabaseManagerClusterRPC(ClientClusterRPC client, Map<ServerAddress, DatabaseManagerRPC> databaseManagers) {
        this.client = client;
        this.databaseManagers = databaseManagers;
    }

    @Override
    public boolean contains(String name) {
        List<GraknClientException> errors = new ArrayList<>();
        for (DatabaseManagerRPC databaseManager : databaseManagers.values()) {
            try {
                return databaseManager.contains(name);
            } catch (GraknClientException e) {
                errors.add(e);
            }
        }
        throw new GraknClientException(CLUSTER_ALL_NODES_FAILED.message(errors.toString()));
    }

    @Override
    public void create(String name) {
        for (DatabaseManagerRPC databaseManager : databaseManagers.values()) {
            if (!databaseManager.contains(name)) {
                databaseManager.create(name);
            }
        }
    }

    @Override
    public GraknClient.Database.Cluster get(String name) {
        List<GraknClientException> errors = new ArrayList<>();
        for (ServerAddress address : databaseManagers.keySet()) {
            try {
                DatabaseProto.Database.Get.Res res = client.graknClusterRPC(address)
                        .databaseGet(DatabaseProto.Database.Get.Req.newBuilder().setName(name).build());
                return DatabaseClusterRPC.of(res.getDatabase(), this);
            } catch (GraknClientException e) {
                errors.add(e);
            }
        }
        throw new GraknClientException(CLUSTER_ALL_NODES_FAILED.message(errors.toString()));
    }

    @Override
    public List<GraknClient.Database.Cluster> all() {
        List<GraknClientException> errors = new ArrayList<>();
        for (ServerAddress address : databaseManagers.keySet()) {
            try {
                DatabaseProto.Database.All.Res res = client.graknClusterRPC(address).databaseAll(DatabaseProto.Database.All.Req.getDefaultInstance());
                return res.getDatabasesList().stream().map(db -> DatabaseClusterRPC.of(db, this)).collect(Collectors.toList());
            } catch (GraknClientException e) {
                errors.add(e);
            }
        }
        throw new GraknClientException(CLUSTER_ALL_NODES_FAILED.message(errors.toString()));
    }

    Map<ServerAddress, DatabaseManagerRPC> databaseManagers() {
        return databaseManagers;
    }
}

File: rpc/cluster/DatabaseManagerClusterRPC.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {
    GraknClient, ClientClusterRPC, DatabaseManagerRPC, GraknClientError, ErrorMessage, DatabaseClusterRPC
} from "../../dependencies_internal";
import CLUSTER_ALL_NODES_FAILED = ErrorMessage.Client.CLUSTER_ALL_NODES_FAILED;
import DatabaseProto from "grakn-protocol/protobuf/cluster/database_pb";

export class DatabaseManagerClusterRPC implements GraknClient.DatabaseManagerCluster {
    private readonly _databaseManagers: {[serverAddress: string]: DatabaseManagerRPC};
    private readonly _client: ClientClusterRPC;

    constructor(client: ClientClusterRPC, databaseManagers: {[serverAddress: string]: DatabaseManagerRPC}) {
        this._client = client;
        this._databaseManagers = databaseManagers;
    }

    async contains(name: string): Promise<boolean> {
        let errors = "";
        for (const address of Object.keys(this._databaseManagers)) {
            try {
                return await this._databaseManagers[address].contains(name);
            } catch (e) {
                errors += `- ${address}: ${e}\n`;
            }
        }
        throw new GraknClientError(CLUSTER_ALL_NODES_FAILED.message(errors));
    }

    async create(name: string): Promise<void> {
        for (const databaseManager of Object.values(this._databaseManagers)) {
            if (!(await databaseManager.contains(name))) {
                await databaseManager.create(name);
            }
        }
    }

    async get(name: string): Promise<DatabaseClusterRPC> {
        let errors = "";
        for (const address of Object.keys(this._databaseManagers)) {
            try {
                const res: DatabaseProto.Database.Get.Res = await new Promise((resolve, reject) => {
                    this._client.graknClusterRPC(address).database_get(new DatabaseProto.Database.Get.Req().setName(name), (err, res) => {
                        if (err) reject(new GraknClientError(err));
                        else resolve(res);
                    });
                });
                return DatabaseClusterRPC.of(res.getDatabase(), this);
            } catch (e) {
                errors += `- ${address}: ${e}\n`;
            }
        }
        throw new GraknClientError(CLUSTER_ALL_NODES_FAILED.message(errors));
    }

    async all(): Promise<DatabaseClusterRPC[]> {
        let errors = "";
        for (const address of Object.keys(this._databaseManagers)) {
            try {
                const res: DatabaseProto.Database.All.Res = await new Promise((resolve, reject) => {
                    this._client.graknClusterRPC(address).database_all(new DatabaseProto.Database.All.Req(), (err, res) => {
                        if (err) reject(new GraknClientError(err));
                        else resolve(res);
                    });
                });
                return res.getDatabasesList().map(db => DatabaseClusterRPC.of(db, this));
            } catch (e) {
                errors += `- ${address}: ${e}\n`;
            }
        }
        throw new GraknClientError(CLUSTER_ALL_NODES_FAILED.message(errors));
    }

    databaseManagers(): {[address: string]: DatabaseManagerRPC} {
        return this._databaseManagers;
    }
}
